src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -5400,16 +5400,21 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
 
           setError(tree)
         }
-          // ignore current variable scope in patterns to enforce linearity
+        // ignore current variable scope in patterns to enforce linearity
         val startContext = if (mode.typingPatternOrTypePat) context.outer else context
 
         def asTypeName = if (mode.inAll(MonoQualifierModes) && unit.isJava && name.isTermName) {
           startContext.lookupSymbol(name.toTypeName, qualifies).symbol
         } else NoSymbol
 
-        val nameLookup   = tree.symbol match {
-          case NoSymbol   => startContext.lookupSymbol(name, qualifies)
-          case sym        => LookupSucceeded(EmptyTree, sym)
+        // in Java, only pick a package if it is rooted
+        def termQualifies(sym: Symbol) = qualifies(sym) && (
+            !startContext.unit.isJava  || !sym.hasPackageFlag
+          || sym.owner.isEffectiveRoot || sym.owner.isRootPackage || sym.isRootPackage
```

Comment:
tempting is the paulpism, `List(sym, sym.owner).exists(s => s.isEffectiveRoot || s.isRootPackage)`

Reply:
Or the introduction of `sym.isRooted`.

Reply:
It actually took me a while (while working on something work-related) to come up with this formulation. I didn't get to the point where I felt comfortable adding `sym.isEffectiveRoot` and I've already forgotten the grid. For `p`, is `p.isEffectiveRoot` or its owner?